### PR TITLE
arch: support arm64

### PR DESCRIPTION
**Pull request by nickolaev**
Adds basic developer support for ARM64, fixed the tests and added some instructions/notes.

Modify `kprobe` to be smart about syscall naming, depending on the platform.
For example, it will take `sys_getpid` and turn it into `__arm64_sys_getpid` when running on an ARM64 host.

Fixes https://github.com/cilium/ebpf/issues/266

Signed-off-by: Nikolay Nikolaev <nicknickolaev@gmail.com>

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,7 +1,14 @@
 # The development version of clang is distributed as the 'clang' binary,
 # while stable/released versions have a version number attached.
 # Pin the default clang to a stable version.
+GOARCH := $(shell go env GOARCH)
+ifneq ($(GOARCH),arm64)
 CLANG ?= clang-12
+else
+# Ubuntu Focal used for the ARM64 builder container does not provide
+# clang-12 out of the box
+CLANG ?= clang-11
```
Comment (lmb) [inline Makefile:10]: Unfortunately we need clang-12 for upcoming CO-RE support, see #301. FWIW I don't think we need to build using arm64 clang, there should be no difference in output for bpfel / bpfeb targets on either platform.

File: examples/kprobe/fn_amd64.go [hunk 2]
```diff
@@ -0,0 +1,4 @@
+package main
+
+// Name of the kernel function to trace.
+const fn = "__x64_sys_execve"
```
Comment (lmb) [inline examples/kprobe/fn_amd64.go:4]: Checking https://golang.org/pkg/runtime/#pkg-constants GOARCH in the code should be simpler.

**Review by lmb — CHANGES_REQUESTED** [review]
Thanks for making the PR :) Please see my comments, I think changing the ELF builder is not necessary. It's nice that we have working tests on arm64, but we can't claim to support arm64 until we have a working CI.

File: examples/kprobe/fn_amd64.go (hunk 2 above)
Comment (nickolaev) [reply examples/kprobe/fn_amd64.go]: OK, changed the approach. Let me know if that works.

File: link/platform.go [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package link
+
+import "runtime"
+
+func byArch(arch_map map[string]string) string {
```
Comment (lmb) [inline link/platform.go:5]: Suggestion: just switch on runtime.GOARCH and prepend `__x64_`, etc. to a string supplied by the user.
```suggestion
func archSyscallName(name string) string {
```

```
archSyscallName("sys_getpid")
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -20,4 +20,32 @@ reason about the proposed changes.
...
+
+## Other platforms
+
+The main development platform that the project targets is `amd64`. However running
+the code and the examples should not be hard on any platform supported by `go` and
+`docker`. Here are some hints for `arm64`, adding other platforms should be similar.
+The suggested approach will always use `amd64` build tools, but allow to cross-build
+and use them with an emulation.
+
+#### Ubuntu `linux/arm64`
```
Comment (lmb) [inline CONTRIBUTING.md:33]: So what commands would you usually run when developing on linux/arm64 then? `go test ./...`? `make`?

File: examples/kprobe/main.go [hunk 5]
```diff
@@ -50,7 +46,7 @@ func main() {
 	// pre-compiled program. Each time the kernel function enters, the program
 	// will increment the execution counter by 1. The read loop below polls this
 	// map value once per second.
-	kp, err := link.Kprobe(fn, objs.KprobeExecve)
+	kp, err := link.Kprobe(link.SysExecve(), objs.KprobeExecve)
```
Comment (lmb) [inline examples/kprobe/main.go:49]: I wonder whether we should move this fallback behaviour into Kprobe, Kretprobe, etc. proper. Otherwise our interface isn't very portable. That is, the user specifies "sys_execve" here, and we discover
* the correct arch prefix
* with a fallback to just the plain name

@ti-mo thoughts?

File: link/platform.go [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+	value, ok := arch_map[runtime.GOARCH]
+	if !ok {
+		panic("Unsupported arch " + runtime.GOARCH)
+	}
+	return value
+}
+
+func SysGetpid() string {
```
Comment (lmb) [inline link/platform.go:13]: I don't think this warrants a stand alone / exported function, see my comments about archSyscallName and maybe baking this into Kprobe, etc.

File: Makefile [hunk 7]
```diff
@@ -31,15 +31,15 @@ TARGETS := \
 
 # Build all ELF binaries using a Dockerized LLVM toolchain.
 docker-all:
-	docker run --rm --user "${UIDGID}" \
+	docker run --platform linux/amd64 --rm --user "${UIDGID}" \
```
Comment (lmb) [inline Makefile:34]: Unfortunately this breaks on my machine:
```
$ make
docker run --platform linux/amd64 --rm --user "1000:1000" \
	-v "/home/lorenz/dev/ebpf":/ebpf -w /ebpf --env MAKEFLAGS \
	--env CFLAGS="-fdebug-prefix-map=/ebpf=." \
	"quay.io/cilium/ebpf-builder:1619686610" \
	make all
"--platform" is only supported on a Docker daemon with experimental features enabled
make: *** [Makefile:34: docker-all] Error 1
```

File: CONTRIBUTING.md (hunk 4 above)
Comment (ti-mo) [inline CONTRIBUTING.md:33]: So Linux == Ubuntu? :) 

If I understand correctly, this section is supposed to describe how to generate a Docker image containing an arm64-based LLVM toolchain. Can't tell for sure because there's no conclusion to the description. This would enable developers running arm64-based Linux distributions as a development environment (which might include Docker-on-M1), is that correct?

If so, I would remove all of the fluff here and make `testdata/docker` rely on buildx. A short mention can be included in the readme there about `qemu-user-static` and `binfmt-support` being required, although this is kind of expected for running a cross-compilation toolchain. `buildx` will likely fail and tell the user these tools are required.

File: Makefile (hunk 7 above)
Comment (ti-mo) [reply Makefile:34]: This change feels a little unnecessary. `docker run` should try to pull and run an image compatible with the host's architecture. (I assume this is the point of distributing arm64 LLVM builder images) Does this actually work on M1? Will it try to pull and run an x86 Docker image and execute it on an emulated x86 host?

If so, what's the point of building arm64 toolchain images? (I assume it's horribly slow for larger eBPF codebases)

File: link/perf_event_test.go [hunk 8]
```diff
@@ -26,7 +26,7 @@ func TestTraceEventTypePMU(t *testing.T) {
 func TestTraceEventID(t *testing.T) {
 	c := qt.New(t)
 
-	eid, err := getTraceEventID("syscalls", "sys_enter_fork")
+	eid, err := getTraceEventID("syscalls", "sys_enter_execve")
```
Comment (ti-mo) [inline link/perf_event_test.go:29]: Why?

File: link/tracepoint_test.go [hunk 9]
```diff
@@ -66,7 +66,7 @@ func TestTracepointErrors(t *testing.T) {
 }
 
 func TestTraceGetEventID(t *testing.T) {
-	_, err := getTraceEventID("syscalls", "sys_enter_open")
+	_, err := getTraceEventID("syscalls", "sys_enter_openat")
```
Comment (ti-mo) [inline link/tracepoint_test.go:69]: Why?

File: examples/kprobe/main.go (hunk 5 above)
Comment (ti-mo) [reply examples/kprobe/main.go:49]: Agreed, this needs to become a part of the API, not a hack in the test suite. That doesn't add much value to the user, as it makes it harder to write tools portable across architectures.

We also need to hash out what the user should expect, as in: do they specify `__x64_sys_execve` and rely on the lib to implicitly fix it up to make it portable? Or do they specify `sys_execve`, and the lib always tries with and without the arch prefix? There might be code out there already specifying `__x64_sys_execve` that will break if we change the rules.

I'd suggest the following:
- If we detect an arch prefix, fix it up to the prefix that belongs to the host. This should work in most cases.
- Then, first, do exactly what the user asked for (`sys_execve`). If the symbol already contains an arch prefix, we can bail out here.
- If that fails, try again with the arch prefix prepended (`__x64_sys_execve`). Syscall symbols are in the minority, so this should avoid the most back-and-forth.

Note: Using tracepoints should be preferred for probing most popular syscalls (execve was only chosen here because it's easy to trigger), but I'm not sure if tracepoints yield the same flexibility in terms of walking memory.

**Review by ti-mo — CHANGES_REQUESTED** [review]
Thanks for the patch! We're seeing arm being used more and more frequently, but I think this needs some more work in order to become useful.

- Please remove all merge commits. `git fetch origin && git rebase origin/master` is the way to keep your branch up-to-date with changes that have landed on master.
- I don't feel like the additions to `CONTRIBUTING.md` are necessary. If we want to provide llvm builder images going forward, `testdata/docker` should output images for both arm64 and x86, which will automatically be picked on hosts/dev machines of those respective architectures.
- `link.K(ret)probe` should handle `__x64_` syscall fn prefixes transparently, but implementing this might be annoying given the difference between tracefs and PMU, see other comment on the code itself. Please reach out through Cilium Slack (`#libbpf-go`) if you need more input.

Marking this as draft in the mean time.

File: examples/kprobe/main.go (hunk 5 above)
Comment (lmb) [reply examples/kprobe/main.go:49]: I think the can do something simpler. Assuming we get `__x64_sys_execve` as the argument:
* First try with the arch specific prefix added: `__x64___x64_sys_execve`. This will fail.
* Fall back to trying without the arch prefix (necessary pre 4.17 anyways): `__x64_sys_execve`. This will work on new kernels.

We shouldn't try to translate arch prefixes from one arch to another. This not any more or less broken than at the moment and a lot simpler.

File: Makefile (hunk 7 above)
Comment (lmb) [reply Makefile:34]: I had a quick look the other day, I'm not sure docker is smart enough unfortunately. Something about us having to publish a manifest. I didn't dig too deep and would love to proven wrong.

File: examples/kprobe/main.go (hunk 5 above)
Comment (ti-mo) [reply examples/kprobe/main.go:49]: You might be right. The user might want to attach only to `__x64_sys_execve` and bail out on other architectures. We also wouldn't have to maintain a list of architecture prefixes to strip, reducing maintenance overhead.

For portable probes, the bare syscall symbol (`sys_execve`) must be specified. This is probably the cleanest API contract.

File: Makefile (hunk 7 above)
Comment (nickolaev) [reply Makefile:34]: > This change feels a little unnecessary. `docker run` should try to pull and run an image compatible with the host's architecture. (I assume this is the point of distributing arm64 LLVM builder images) Does this actually work on M1? Will it try to pull and run an x86 Docker image and execute it on an emulated x86 host?
> 
> If so, what's the point of building arm64 toolchain images? (I assume it's horribly slow for larger eBPF codebases)

I have tried this one on my M1 in both Docker Desktop for Mac and my Parallels Ubuntu 20.04 VM. Both work with this flag, and I would say is usable for the needs of the development within this project. At this point, I don't think there is any reason to build arm64 images at all.

File: link/perf_event_test.go (hunk 8 above)
Comment (nickolaev) [reply link/perf_event_test.go:29]: From what I can tell `sys_enter_fork` was not available on arm, so I thought that for the purposes of testing `sys_enter_execve` is more portable.

File: link/tracepoint_test.go (hunk 9 above)
Comment (nickolaev) [reply link/tracepoint_test.go:69]: Same as above, just a more widely implemented syscall, that is available on other archs.

**Comment by nickolaev** [comment]
@lmb and @ti-mo, I've tried to be creative here and figure a better way to handle syscall naming cross platforms and kernels. Let me know if that aligns with your ideas here.

File: internal/kernel.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
+package internal
```
Comment (lmb) [inline internal/kernel.go:1]: check out internal/version.go

File: link/platform.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+	"regexp"
+	"runtime"
+
+	"github.com/cilium/ebpf/internal"
+)
+
+const (
+	// Since 4.17, syscalls symbols are generated with the `__x64_` prefix.
+	// https://github.com/torvalds/linux/commit/d5a00528b58cdb2c71206e18bd021e34c4eab878
+	LinuxVersionWithPrefix = 4<<16 + 17<<8 + 0
```
Comment (lmb) [inline link/platform.go:15]: Version comparisons are brittle, since individual distributions freeze or change version numbers according to their own whims. It's always preferable to attempt an action and fall back to old behaviour if necessary.

File: link/platform.go [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+)
+
+var parseArchSyscall = regexp.MustCompile(`__(\w*)_sys_(\w*)`)
+var parseSyscall = regexp.MustCompile(`sys_(\w*)`)
+
+func ResolveSyscall(syscall string) (string, error) {
```
Comment (lmb) [inline link/platform.go:21]: Why do you want to support this API? The user should give us the "platform independent" syscall name, then we don't need all of this code.

**Review by lmb — CHANGES_REQUESTED** [review]
Please take a look at https://github.com/cilium/ebpf/pull/304#discussion_r636994287 which outlines a simpler API.

File: examples/go.mod [hunk 13]
```diff
@@ -6,3 +6,5 @@ require (
 	github.com/cilium/ebpf v0.5.1-0.20210428090942-9aea0e1002a2
 	golang.org/x/sys v0.0.0-20210124154548-22da62e12c0c
 )
+
+replace github.com/cilium/ebpf => ./..
```
Comment (lmb) [inline examples/go.mod:10]: I don't think this will work: https://github.com/golang/go/issues/30354 The way to go about this is to make the change to examples in a separate PR, after this change has been merged. Otherwise there is no way to update the dependency.

File: link/platform.go [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+
+import (
+	"fmt"
+	"runtime"
+)
+
+func platformPrefix(symbol string) string {
+
+	prefix := runtime.GOARCH
+	switch prefix {
```
Comment (lmb) [inline link/platform.go:11]: Do you have a reference / link which prefixes the kernel uses? Your code currently assumes that GOARCH == kernel prefix in general, I'm not sure that will be true.

**Review by lmb — CHANGES_REQUESTED** [review]

File: link/platform.go (hunk 14 above)
Comment (ti-mo) [reply link/platform.go:11]: > Your code currently assumes that GOARCH == kernel prefix in general, I'm not sure that will be true.

Out of curiosity, what do you mean by this? I fail to come up with an example. Perhaps when running the binary under `qemu-user-static`? What other facilities would the runtime expose to detect underlying OS?

File: link/platform.go [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+	case "i386":
```
Comment (ti-mo) [inline link/platform.go:12]: This is not a valid GOARCH value. All possible values are here: https://github.com/golang/go/blob/master/src/go/build/syslist.go.

File: link/platform.go (hunk 14 above)
Comment (ti-mo) [reply link/platform.go:11]: > Do you have a reference / link which prefixes the kernel uses?

Tricky, but from what I can find:

```
# in linux
λ  linux master ✗ egrep -R "define __SYSCALL(.*) = __" arch
arch/x86/entry/syscall_64.c:#define __SYSCALL_64(nr, sym) [nr] = __x64_##sym,
arch/x86/entry/syscall_32.c:#define __SYSCALL_I386(nr, sym) [nr] = __ia32_##sym,
arch/x86/entry/syscall_x32.c:#define __SYSCALL_X32(nr, sym) [nr] = __x32_##sym,
arch/x86/entry/syscall_x32.c:#define __SYSCALL_COMMON(nr, sym) [nr] = __x64_##sym,
arch/arm64/kernel/sys32.c:#define __SYSCALL(nr, sym)	[nr] = __arm64_##sym,
arch/arm64/kernel/sys.c:#define __SYSCALL(nr, sym)	[nr] = __arm64_##sym,
```

Comment (lmb) [reply link/platform.go:11]: > Out of curiosity, what do you mean by this?

The code relies on the fact that on arm64 GOARCH == the prefix used by the kernel, and assumes that the same is true for other platforms. Nothing to do with changes in GOARCH due to different runtimes.

File: link/platform.go [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+		prefix = "ia32"
+	case "amd64":
+		prefix = "x64"
+	}
```
Comment (ti-mo) [inline link/platform.go:16]: This branch/PR is targeting arm64, but I don't see an arm64 entry here. :) Note that according to https://github.com/golang/go/blob/master/src/go/build/syslist.go there's also an `arm64be` entry, but the Go compiler doesn't seem to support that out of the box:
```
λ  ~/documents/go/sandbox  GOARCH=arm64be go build
cmd/go: unsupported GOOS/GOARCH pair linux/arm64be
```
so I think we can skip an arm64be entry.

File: link/platform.go (hunk 14 above)
Comment (ti-mo) [reply link/platform.go:11]: Oh, completely missed that. We should probably return empty by default and only emit known prefixes.

@iii-i (sorry for the random ping), could you chime in on what's common in s390(x)? I'm finding references to both `__s390_` and `__s390x_` in Linux.

Comment (nickolaev) [reply link/platform.go:11]: I assume we can cover as much as possible of the options. I tried to cover x86 and arm64. We can panic on unknown platform and let these be added over time once users hit "unsupported platform".

File: link/platform.go (hunk 15 above)
Comment (nickolaev) [reply link/platform.go:12]: True, my eyes misled me on this. From what I recall I checked this gist -> https://gist.github.com/asukakenji/f15ba7e588ac42795f421b48b8aede63#a-list-of-valid-goarch-values

File: link/platform.go (hunk 16 above)
Comment (nickolaev) [reply link/platform.go:16]: Yes, my approach was to "patch" only for those platform that differ from GOARCH. Maybe it is better to make these translations more explicint in map?

File: examples/go.mod (hunk 13 above)
Comment (nickolaev) [reply examples/go.mod:10]: It depends on what is `works`. For sure with this patch, I am able to run the kprobe example and am not able to do so without it.

File: link/platform.go (hunk 14 above)
Comment (ti-mo) [reply link/platform.go:11]: As this is library code, we cannot panic. The safer option would be to not return a prefix, so the given symbol is plumbed through as is. Of course, we'd want to ship this with most of the available architectures supported as you suggested..

File: README.md [hunk 17]
```diff
@@ -57,6 +57,12 @@ environment to keep the build output stable.
 
 The toolchain image build files are kept in [testdata/docker/](testdata/docker/).
 
+### Regeneration Testdata on non-x86 platforms
```
Comment (ti-mo) [inline README.md:60]: Could this be moved to `testdata/docker/README.md` under the `Building` section?

File: link/platform.go (hunk 16 above)
Comment (ti-mo) [reply link/platform.go:16]: The structure is fine as you suggested, I would just add a default case with `return symbol`.

**Review by lmb — APPROVED** [review]

**Comment by ti-mo** [comment]
@nickolaev Thanks for the work!
